pkg/validators/kubeletconfigvalidator.go

```diff
@@ -0,0 +1,211 @@
+		if err != nil {
+			return nil, err
+		}
+
+		machineConfigPools, err := kubeletconfig.GetMCPsFromMCOKubeletConfig(gatherer.conn.context, gatherer.conn.client, mcoKubeletConfig)
+		if err != nil {
+			return nil, err
+		}
+		for _, mcp := range machineConfigPools {
+
+			nodes, err := machineconfigpools.GetNodeListFromMachineConfigPool(gatherer.conn.context, gatherer.conn.client, *mcp)
+			if err != nil {
+				return nil, err
+			}
+			nodeNames := getNodeNames(nodes)
+
+			for _, nodeName := range nodeNames {
+				// we are just interested on nodes with TAS enabled
+				if !data.tasEnabledNodeNames.Has(nodeName) {
+					continue
+				}
+
+				_, found := data.kConfigs[nodeName]
+				if found {
+					return nil, fmt.Errorf("Found two KubeletConfigurations for node %q", nodeName)
+				}
+				data.kConfigs[nodeName] = kubeletConfig
+			}
+		}
+	}
+
+	// get Version Info
+	ver, err := getClusterVersionInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	data.versionInfo = ver
+	return data, nil
+}
+
+type KubeletConfigValidator struct {
+	connection   k8sConnection
+	data         kubeletConfigValidatorData
+	dataGatherer KubeletConfigDataGatherer
+}
+type k8sConnection struct {
+	client  client.Client
+	context context.Context
+}
+type kubeletConfigValidatorData struct {
+	tasEnabledNodeNames sets.String
+	kConfigs            map[string]*kubeletconfigv1beta1.KubeletConfiguration
+	versionInfo         *version.Info
+}
+
+func NewKubeletConfigValidator(aContext context.Context, aClient client.Client) *KubeletConfigValidator {
+	defaultGatherer := &k8KubeletConfigDataGatherer{
+		conn: k8sConnection{
+			client:  aClient,
+			context: aContext,
+		},
+		nroName: controllers.DefaultNUMAResourcesOperatorCrName,
+	}
+	return NewKubeletConfigValidatorWithDataGatherer(aContext, aClient, defaultGatherer)
+}
+
+func NewKubeletConfigValidatorWithDataGatherer(aContext context.Context, aClient client.Client, aDataGatherer KubeletConfigDataGatherer) *KubeletConfigValidator {
+
+	ret := &KubeletConfigValidator{
+		connection: k8sConnection{
+			context: aContext,
+			client:  aClient,
+		},
+		dataGatherer: aDataGatherer,
+	}
+	return ret
+}
+
+func (kvalidator *KubeletConfigValidator) Setup() error {
+
+	data, err := kvalidator.dataGatherer.CollecData()
+	if err != nil {
+		return err
+	}
+	kvalidator.data = *data
+	return nil
+}
+
+func (kvalidator *KubeletConfigValidator) Validate() ([]validator.ValidationResult, error) {
+
+	var ret []validator.ValidationResult
+	for _, nodeName := range kvalidator.data.tasEnabledNodeNames.List() {
+		kc := kvalidator.data.kConfigs[nodeName]
+		// if a TAS enabled node has no MCO kubelet configuration kc will be nil and ValidateClusterNodeKubeletConfig will fill the proper error
+		res := validator.ValidateClusterNodeKubeletConfig(nodeName, kvalidator.data.versionInfo, kc)
+		ret = append(ret, res...)
+	}
+	return ret, nil
+}
+
+func (kvalidator *KubeletConfigValidator) Teardown() error {
+	// nothing to do
```

Comment:
then why we need this abstraction? it also seems this is the only implementation of the aforementioned abstraction

Reply:
Yep, right now this is the only implementation. 
I thought we will probable need to validate some other things apart from the kubeletconfig and having this abstraction could help us to separate the creation of the validators from the execution of the validation

Reply:
The intention is good, but I'd rather build a interface when use cases are there and a bit more defined.